### PR TITLE
bpo-45613: Set `sqlite3.threadsafety` dynamically

### PR DESCRIPTION
**Pull request by erlend-aasland**
Use the compile-time selected default SQLite threaded mode to set the
DB-API 2.0 attribute 'threadsafety' during sqlite3 module initialisation.

SQLite to DB-API mappings:
  - SQLITE_THREADSAFE=0 => threadsafety=0 (single-thread mode)
  - SQLITE_THREADSAFE=1 => threadsafety=3 (fully threadsafe)
  - SQLITE_THREADSAFE=2 => threadsafety=1 (threadsafe at module level)

https://bugs.python.org/issue45613

### CODE REVIEW

File: Modules/_sqlite/module.c [hunk 1]
```diff
@@ -415,6 +415,28 @@ add_integer_constants(PyObject *module) {
...
+        return 0;
+    case 1:        // Serialized mode; threads may share the module,
+        return 3;  // connections, and cursors.
+    case 2:        // Multi-thread mode; threads may share the module, but not
+        return 1;  // connections.
+    default:
+        PyErr_Format(state->InterfaceError,
+                     "Unable to interpret SQLite threadsafety mode. Got %d, "
+                     "expected 0, 1, or 2", mode);
+        return -1;
```
Comment (erlend-aasland) [inline Modules/_sqlite/module.c:436]: Maybe it is better to default to multi-thread mode (threadsafety 1), possibly raising a warning.

File: Doc/library/sqlite3.rst [hunk 2]
```diff
@@ -159,22 +159,38 @@ Module functions and constants
...
+   +------------------+-----------------+----------------------+-------------------------------+
+   | multi-thread     | 1               | 2                    | Threads may share the module, |
+   |                  |                 |                      | but not connections           |
+   +------------------+-----------------+----------------------+-------------------------------+
+   | serialized       | 3               | 1                    | Threads may share the module, |
+   |                  |                 |                      | connections and cursors       |
+   +------------------+-----------------+----------------------+-------------------------------+
+
+   .. _threadsafety: https://www.python.org/dev/peps/pep-0249/#threadsafety
+   .. _SQLITE_THREADSAFE: https://sqlite.org/compile.html#threadsafe
```
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst:193]: KISS version:

```suggestion
underlying SQLite library is compiled with.
```

Comment (malemburg) [reply Doc/library/sqlite3.rst:193]: I think more context is always good for these more complex things, so would recommend leaving the full version.

File: Modules/_sqlite/module.c (hunk 1 above)
Comment (malemburg) [reply Modules/_sqlite/module.c:436]: No. If SQLite returns an unexpected value, it's better to raise an error.

**Review by malemburg — APPROVED** [review]
I think the patch is good to go as-is, without the additional changes you mentioned in the comments. Thanks, Erlend.

**Comment by erlend-aasland** [comment]
> I think the patch is good to go as-is, without the additional changes you mentioned in the comments. Thanks, Erlend.

Thanks! There is one change I'd like to make before merging, though: add a `versionchanged` to the `threadsafety` attribute doc.

**Review by pablogsal — APPROVED** [review]
LGTM

@malemburg do you have anything else you would like here to be addressed?

Edit: ah, I just saw that you approved before. I am landing this then

**Comment by erlend-aasland** [comment]
Thanks, Marc-Andre & Pablo 🙏🏻

**Pull request merged**